android/expoview/src/main/java/host/exp/exponent/ExpoUpdatesAppLoader.java

```diff
@@ -214,14 +220,23 @@ public void onSuccess(Launcher launcher) {
       }
 
       @Override
-      public void onEvent(String eventName, WritableMap params) {
+      public void onBackgroundUpdateFinished(LoaderTask.BackgroundUpdateStatus status, @Nullable UpdateEntity update, @Nullable Exception exception) {
         try {
           JSONObject jsonParams = new JSONObject();
-          jsonParams.put("type", eventName);
-          Iterator<Map.Entry<String, Object>> iterator = params.getEntryIterator();
-          while (iterator.hasNext()) {
-            Map.Entry<String, Object> entry = iterator.next();
-            jsonParams.put(entry.getKey(), entry.getValue());
+          if (status == LoaderTask.BackgroundUpdateStatus.ERROR) {
+            if (exception == null) {
+              throw new AssertionError("Background update with error status must have a nonnull exception object");
+            }
+            jsonParams.put("type", UPDATE_ERROR_EVENT);
+            jsonParams.put("message", exception.getMessage());
+          } else if (status == LoaderTask.BackgroundUpdateStatus.UPDATE_AVAILABLE) {
+            if (update == null) {
+              throw new AssertionError("Background update with error status must have a nonnull update object");
+            }
+            jsonParams.put("type", UPDATE_AVAILABLE_EVENT);
+            jsonParams.put("manifestString", update.metadata.toString());
+          } else if (status == LoaderTask.BackgroundUpdateStatus.NO_UPDATE_AVAILABLE) {
+            jsonParams.put("type", UPDATE_NO_UPDATE_AVAILABLE_EVENT);
           }
```

Comment:
A nice thing about this refactoring is that you can see the fields of the different event objects more readily.